### PR TITLE
Dispatcher groups limits

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR exposes new alertmanager limit: number of active dispatcher groups in Alertmanager.

This PR builds on top of #4237, and will be rebased once #4237 is merged.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
This PR builds on top of #4237, marking as draft until it is merged.

**Comment by pstibrany** [comment]
Rebased on top of master now, ready for review.

File: pkg/alertmanager/alertmanager_test.go
Comment (stevesg) [inline pkg/alertmanager/alertmanager_test.go:85]: TIL :)

File: pkg/alertmanager/alertmanager_test.go
Comment (stevesg) [inline pkg/alertmanager/alertmanager_test.go:71]: (non blocking nit) I was thinking whether we should we test the case of having multiple alerts per group? i.e. This particular limit shouldn't be hit with a much higher number of alerts than groups. Maybe we create 2x alerts per group?

**Review by stevesg — APPROVED** [review]
Can't see anything wrong here!

File: pkg/alertmanager/alertmanager_test.go
Comment (pstibrany) [reply pkg/alertmanager/alertmanager_test.go:71]: Sounds good, I'll do that.

File: pkg/util/validation/limits.go
Comment (pracucci) [inline pkg/util/validation/limits.go:185]: We're not used to add "number of" to config fields / CLI flags (most limits are a limit on "number of <something>"). I'm wondering if we could simplify it just calling it `alertmanager.max-dispatcher-aggregation-groups` (and same for YAML/JSON).

**Review by pracucci — COMMENTED** [review]
Good job! LGTM (modulo a comment about config name which I think we can make a bit more consistent with the rest of the config).

Should we mark these new limits as experimental?

**Review by pracucci — APPROVED** [review]

File: pkg/util/validation/limits.go
Comment (pstibrany) [reply pkg/util/validation/limits.go:185]: Sounds good to me.

**Comment by pstibrany** [comment]
Thanks for review.

> Should we mark these new limits as experimental?

I've marked all recently added Alertmanager limits as experimental for now.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:41]:
```suggestion
* [ENHANCEMENT] Alertmanager: Added `-alertmanager.max-dispatcher-aggregation-groups` option to control max number of active dispatcher groups in Alertmanager (per tenant, also overrideable). When the limit is reached, Dispatcher produces log message and increases `cortex_alertmanager_dispatcher_aggregation_group_limit_reached_total` metric. #4254
```

**Review by pracucci — APPROVED** [review]
LGTM (modulo a fix to the CHANGELOG entry). Thanks!

**Pull request merged**
